### PR TITLE
feat(koa): add requestHook support

### PR DESCRIPTION
**Pull request by luismiramirez**
The `requestHook` support allows custom span handling per middleware
layer on requests.

## Which problem is this PR solving?

- Adds `requestHook` support so users can add custom handling on spans per middleware layer on requests.

## Short description of the changes

- It's inspired by how `requestHook` is implemented in other instrumentations
- The `requestHook` user-defined function is called any time a patched middleware runs and if the function exists.

## Checklist

- [x] Ran `npm run test-all-versions` for the edited package(s) on the latest commit if applicable.

### CODE REVIEW

File: plugins/node/opentelemetry-instrumentation-koa/src/types.ts [hunk 1]
```diff
@@ -31,12 +32,28 @@ export type KoaMiddleware = Middleware<DefaultState, KoaContext> & {
 
 export type KoaContext = ParameterizedContext<DefaultState, RouterParamContext>;
 
+export type KoaRequestInfo = {
+  context: KoaContext;
+  middlewareLayer: KoaMiddleware;
+};
+
```
[the remark this answers is not in the document]
Comment (luismiramirez) [reply plugins/node/opentelemetry-instrumentation-koa/src/types.ts:39]: Thanks for noticing. This should be part of the `requestHook`. Fixing it.

[the remark this answers is not in the document]
Comment (luismiramirez) [reply plugins/node/opentelemetry-instrumentation-koa/src/types.ts:39]: Fixed! The idea is to replicate the behaviour suggested here: https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1091#discussion_r923425157

[the remark this answers is not in the document]
Comment (osherv) [reply plugins/node/opentelemetry-instrumentation-koa/src/types.ts:39]: Oh yea, i wasn't sure if that's what u meant to do.
Nice :)

**Review by osherv — APPROVED** [review]

**Review by vmarchaud — APPROVED** [review]

File: plugins/node/opentelemetry-instrumentation-koa/test/koa.test.ts [hunk 2]
```diff
@@ -459,6 +460,112 @@ describe('Koa Instrumentation', () => {
...
+          setRPCMetadata(
+            trace.setSpan(context.active(), rootSpan),
+            rpcMetadata
+          ),
+          next
+        )
+      );
+
+      const requestHook = sinon.spy((span: Span, info: KoaRequestInfo) => {
+        span.setAttribute('http.method', info.context.request.method);
```
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-koa/test/koa.test.ts:478]: better to use the enum `SemanticAttributes.HTTP_METHOD`

**Review by blumamir — APPROVED** [review]

**Comment by vmarchaud** [comment]
@luismiramirez friendly ping because the lint is failling

**Comment by luismiramirez** [comment]
@vmarchaud Linter issues addressed!

**Pull request merged**